src/vec.rs

```diff
@@ -244,12 +244,27 @@ macro_rules! impl_vecn_float_methods {
 
         /// Returns `self` normalized to length 1.0.
         ///
-        /// For valid results, `self` must _not_ be of length zero.
+        /// For valid results, `self` must _not_ be of length zero, nor very close to zero.
+        /// You can check the results with [`Self::is_finite`], or use [`Self::normalize_or_zero`].
         #[inline(always)]
         pub fn normalize(self) -> Self {
             Self($flttrait::normalize(self.0))
         }
 
+        /// Returns `self` normalized to length 1.0 if possible, else returns zero.
+        ///
+        /// In particular, if the input is zero (or very close to zero), or non-finite,
+        /// the result of this operation will be zero.
+        #[inline(always)]
+        pub fn normalize_or_zero(self) -> Self {
+            let v = self.normalize();
+            if v.is_finite() {
```

Comment:
You could probably make this cheaper by checking if the reciprocal length is finite rather than every element of the resulting vector. The default trait implementation of normalize is:

```
fn normalize(self) -> Self {
    self.mul_scalar(self.length_recip())
}
```
Only checking rcp would be.
```
fn normalize_or_zero(self) -> Self {
     let rcp = self.length_recip();
     if rcp.is_finite() {
         self.mul_scalar(rcp)
    } else {
        Self::ZERO
    }
}
```
Perhaps that could still miss degenerate results though?

Reply:
I expected it to fail for an infinite vector, but surprisingly `Vec2::new(INFINITY, 0.0).length()` is `NAN` rather than `INFINITY`, and `Vec2::new(INFINITY, 0.0).length_recip()` is `NAN` rather than zero. I'm not sure if that will be true on all platforms though (maybe it is a simd thing?).

`if rcp.is_finite() && rcp > 0.0` may be a safer test

Reply:
Nevermind that comment - my tests where wrong. `Vec2::new(INFINITY, 0.0).length()` is INFINITY.

The `if rcp.is_finite()` version does NOT work for infinite vectors, but `if rcp.is_finite() && rcp > 0.0` works for all my tests, and is faster than my original version. On my machine the original `normalize + is_finite` version was 80% slower than vanilla `normalize`, but the new `rcp.is_finite() && rcp > 0.0` is only around 40% slower than vanilla `normalize`